crates/aptos-logger/src/speculative_log.rs

```diff
@@ -0,0 +1,284 @@
+    base_version: Option<Version>,
+
+    // transaction position in the list of transactions in the block,
+    // 0 if the transaction is not part of a block (i.e. validation).
+    txn_idx: usize,
+}
+
+impl AdapterLogSchema {
+    pub fn new(view_id: StateViewId, txn_idx: usize) -> Self {
+        match view_id {
+            StateViewId::BlockExecution { block_id } => Self {
+                name: LogEntryKind::Execution,
+                block_id: Some(block_id),
+                first_version: None,
+                base_version: None,
+                txn_idx,
+            },
+            StateViewId::ChunkExecution { first_version } => Self {
+                name: LogEntryKind::Execution,
+                block_id: None,
+                first_version: Some(first_version),
+                base_version: None,
+                txn_idx,
+            },
+            StateViewId::TransactionValidation { base_version } => Self {
+                name: LogEntryKind::Validation,
+                block_id: None,
+                first_version: None,
+                base_version: Some(base_version),
+                txn_idx,
+            },
+            StateViewId::Miscellaneous => Self {
+                name: LogEntryKind::Miscellaneous,
+                block_id: None,
+                first_version: None,
+                base_version: None,
+                txn_idx,
+            },
+        }
+    }
+}
+
+#[derive(Clone, Copy, Serialize)]
+#[serde(rename_all = "snake_case")]
+pub enum LogEntryKind {
+    Execution,
+    Validation,
+    Miscellaneous, // usually testing
+}
+
+#[macro_export]
+macro_rules! speculative_log {
+    ($level:expr, $($args:tt)+) => {
+        if enabled!($level) {
+            speculative_log($level, $($args)+);
+        }
+    };
+}
+
+// Logs directly (no buffering).
+fn log_message(level: Level, context: &AdapterLogSchema, message: &String) {
+    match level {
+        Level::Error => error!(*context, "{}", message),
+        Level::Warn => warn!(*context, "{}", message),
+        Level::Info => info!(*context, "{}", message),
+        Level::Debug => debug!(*context, "{}", message),
+        Level::Trace => trace!(*context, "{}", message),
+    }
+}
+
+/// Adds a message at a specified logging level and given context (that includes txn index)
+/// to the ongoing buffer.
+/// Note: flush_speculativelogs should be called earlier with the appropriate size to
+/// make sure the underlying BUFFERED_LOGS storage is indexed sufficiently.
+pub fn speculative_log(level: Level, context: &AdapterLogSchema, message: String, critical: bool) {
+    let context = context.clone();
+
+    let txn_idx = context.txn_idx;
+    let logs = BUFFERED_LOGS.load();
+    let mut log = logs.0[txn_idx].lock();
+
+    // Be defensive if e.g. speculative_log is called before initialization. Can pass the log
+    // through but also log a critical error.
+    if txn_idx >= logs.0.len() {
```

Comment:
Under which scenario can this happen?

Reply:
Hypothetically if somehow we executed via aptos-vm but it didn't hit the execute_block method that resizes the buffer properly. It should never happen, but if I missed something and in general, protecting against a panic.

Reply:
oh but the comment is outdated, will update

Reply:
Then should this check come before line 133?

Reply:
VoteProposalSignatureNotFound

Reply:
of course, thanks!